File: python/physical_printer.py
import logging
import os
import threading

from printing import PhysicalPrinter, PrintQueue, PrinterDriverFactory, PrintMetadata, Toner


# This is a class you'd like to get under test so you can change it safely.
class DiagramPhysicalPrinter:
    def __init__(self, physical_printer: PhysicalPrinter = None, print_queue: PrintQueue = None):
        self._physical_printer = physical_printer or PhysicalPrinter()
        self._print_queue = print_queue or PrintQueue(self._physical_printer)
        self._logger = logging.getLogger("DiagramPhysicalPrinter")

    def do_print(self, printable_diagram, info, target_filename):
        printer_driver = PrinterDriverFactory.get_instance().create_driver_for_print()
        printer_driver.set_diagram(printable_diagram.diagram)

        data = PrintMetadata(info.file_type)
        mutex = threading.Lock()
        success = False

        with ((mutex)):
            if not self._physical_printer.is_available or not (self._physical_printer.toner_levels[Toner.Black] > 0 and
                                                               self._physical_printer.toner_levels[Toner.Cyan] > 0 and
                                                               self._physical_printer.toner_levels[Toner.Magenta] > 0 and
                                                               self._physical_printer.toner_levels[Toner.Yellow] > 0):
                self._logger.info("Physical Printer Unavailable")
            elif self._physical_printer.job_count < 0:
                self._logger.info("Physical Printer Unavailable Due to Job Count Inconsistency")
            else:
                # Print the diagram using the Physical Printer
                self._print_queue.add(data)
                summary_information = printable_diagram.summary_information()
                self._logger.info(f"Diagram Summary Information {summary_information}")

                is_summary = len(summary_information) > 10
                if self._physical_printer.start_document(not is_summary, False, "DiagramPhysicalPrinter"):
                    if printer_driver.print_to(self._physical_printer):
                        self._logger.info("Physical Printer Successfully printed")
                        success = True
                    self._physical_printer.end_document()

                if success:
                    # save a backup of the printed document as pdf
                    if os.path.exists(data.filename):
                        self._logger.info(f"Saving backup of printed document as PDF to file {target_filename}")
                        printable_diagram.print_to_file(data.filename, target_filename)

        printer_driver.release_diagram()
        return success

File: typescript/src/physical_printer.ts
import * as fs from 'fs';
import {FlowchartDiagram, IDiagramMetadata} from './documents';
import {PhysicalPrinter, PrinterDriverFactory, PrintMetadata, PrintQueue, Toner} from './printing';
import {Mutex} from 'async-mutex';
import {PrintableDiagram} from "./printableDiagram"; // for thread-safe locking

class DiagramPhysicalPrinter {
  private _physicalPrinter: PhysicalPrinter;
  private _printQueue: PrintQueue;

  constructor(physicalPrinter?: PhysicalPrinter, printQueue?: PrintQueue) {
    this._physicalPrinter = physicalPrinter || new PhysicalPrinter();
    this._printQueue = printQueue || new PrintQueue(this._physicalPrinter);
  }

  async doPrint(
    diagram: PrintableDiagram,
    info: IDiagramMetadata,
    targetFilename: string
  ): Promise<boolean> {
    const printerDriver = PrinterDriverFactory.getInstance().createDriverForPrint();
    printerDriver.setDiagram(diagram.getDiagram() as FlowchartDiagram);

    const data = new PrintMetadata(info.fileType);
    const mutex = new Mutex();
    let success = false;

    try {
      await mutex.acquire();

      const hasEnoughToner = this._physicalPrinter.tonerLevels.get(Toner.Black)! > 0 &&
        this._physicalPrinter.tonerLevels.get(Toner.Cyan)! > 0 &&
        this._physicalPrinter.tonerLevels.get(Toner.Magenta)! > 0 &&
        this._physicalPrinter.tonerLevels.get(Toner.Yellow)! > 0;

      if (!this._physicalPrinter.isAvailable || !hasEnoughToner) {
        console.info("Physical Printer Unavailable");
      } else if (this._physicalPrinter.jobCount < 0) {
        console.info("Physical Printer Unavailable Due to Job Count Inconsistency");
      } else {
        // Print the diagram using the Physical Printer
        this._printQueue.add(data);
        const summaryInformation = diagram.summaryInformation();
        console.info(`Diagram Summary Information ${summaryInformation}`);

        const isSummary = summaryInformation.length > 10;
        if (this._physicalPrinter.startDocument(!isSummary, false, "DiagramPhysicalPrinter")) {
          if (printerDriver.printTo(this._physicalPrinter)) {
            console.info("Physical Printer Successfully printed");
            success = true;
          }
          this._physicalPrinter.endDocument();
        }

        if (success) {
          // save a backup of the printed document as pdf
          if (fs.existsSync(data.filename)) {
            console.info(`Saving backup of printed document as PDF to file ${targetFilename}`);
            diagram.printToFile(data.filename, targetFilename);
          }
        }
      }
    } finally {
      mutex.release();
      printerDriver.releaseDiagram();
    }

    return success;
  }
}

export { DiagramPhysicalPrinter };
